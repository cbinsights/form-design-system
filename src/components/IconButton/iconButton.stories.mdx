import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs, text, boolean, optionsKnob as options } from '@storybook/addon-knobs';
import StarFilledIcon from 'lib/icons/react/StarFilledIcon';
import CollectionPublicIcon from 'lib/icons/react/CollectionPublicIcon';
import StoryAddIcon from 'lib/icons/react/StoryAddIcon';
import WorkFasterIcon from 'lib/icons/react/WorkFasterIcon';
import SalesforceIcon from 'lib/icons/react/SalesforceIcon';
import { arrayToOptions, StoryWrapper, StoryItem, ImportPath, State } from "components/util/storybook";
import { MockLink } from 'components/util/mock-react-router';

import IconButton, { THEMES, RADII, SIZES } from '.';

export const Icons = [StarFilledIcon, CollectionPublicIcon, StoryAddIcon, WorkFasterIcon, SalesforceIcon]

<Meta title="Components/IconButton" decorators={[withKnobs]} component={IconButton} />

# IconButton

<ImportPath component={IconButton} section="interactive" />

<Story name="Knobs" parameters={{ docs: { disable: true }}}>
  <div className={boolean("Inverted", false) ? 'inverted' : ''}>
    <IconButton
      isDestructive={boolean('isDestructive', false)}
      isActive={boolean('isActive', false)}
      disabled={boolean('disabled', false)}
      isLoading={boolean('isLoading', false)}
      theme={options('theme', arrayToOptions(THEMES), undefined, {
        display: 'inline-radio',
      })}
      size={options('size', arrayToOptions(SIZES), undefined, {
        display: 'inline-radio',
      })}
      radius={options('radius', arrayToOptions(RADII), undefined, {
        display: 'inline-radio',
      })}
      label={text('label', undefined)}
      Icon={StarFilledIcon}
    />
  </div>
</Story>

<Preview isExpanded>
  <IconButton
    Icon={StarFilledIcon}
  />
</Preview>

<Props of={IconButton} />

### Themes

<Preview>
  <Story name="Themes">
    <StoryWrapper>
      {THEMES.map(theme => (
        <StoryItem>
          <IconButton
            theme={theme}
            Icon={Icons[0]}
          />
        </StoryItem>
      ))}
    </StoryWrapper>
  </Story>
</Preview>

### Destructive

<Preview>
  <Story name="Destructive">
    <StoryWrapper>
      <StoryItem>
        <IconButton
          isDestructive
          Icon={Icons[0]}
      />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>


### Inverted

<Preview>
  <Story name="Inverted">
    <div className="inverted bgColor--blue padding--all rounded--all">
    <StoryWrapper>
      <StoryItem>
        <IconButton
          theme="ghost"
          Icon={Icons[0]}
      />
      </StoryItem>
      <StoryItem>
        <IconButton
          theme="outlined"
          Icon={Icons[0]}
      />
      </StoryItem>
    </StoryWrapper>
    </div>
  </Story>
</Preview>

### isActive

`isActive` should be used on buttons to signify the button having
some affect (applying a filter, causing a popover to be open).
**Click to toggle on and off**.

<Preview>
  <Story name="isActive">
    <>
      <StoryWrapper>
        <State
          initialValue={{}}
          render={(value, setValue) => ( 
            <>
              {THEMES.map((theme, idx) => (
                <StoryItem>
                  <IconButton
                    Icon={Icons[0]}
                    theme={theme}
                    isActive={!value[idx]}
                    onClick={() => setValue({ ...value, [idx]: !Boolean(value[idx])})}
                  />
                </StoryItem>
              ))}
            </>
          )}
        />
        <State
          initialValue={{}}
          render={(value, setValue) => ( 
            <>
              <StoryItem>
                <IconButton
                  isDestructive
                  Icon={Icons[0]}
                  isActive={!value[0]}
                  onClick={() => setValue({ ...value, [0]: !Boolean(value[0])})}
                />
              </StoryItem>
            </>
          )}
        />
      </StoryWrapper>
      <div class="inverted bgColor--blue padding--all rounded--all margin--top">
        <StoryWrapper>
          <State
            initialValue={{}}
            render={(value, setValue) => ( 
              <>
                {['ghost', 'outlined'].map((theme, idx) => (
                  <StoryItem>
                    <IconButton
                      Icon={Icons[0]}
                      theme={theme}
                      isActive={!value[idx]}
                      onClick={() => setValue({ ...value, [idx]: !Boolean(value[idx])})}
                    />
                  </StoryItem>
                ))}
              </>
            )}
          />
        </StoryWrapper>
      </div>
    </>
  </Story>
</Preview>


### Sizes

<Preview>
  <Story name="Sizes">
    <StoryWrapper>
    {SIZES.map(size => (
      <StoryItem>
        <IconButton
          size={size}
          Icon={Icons[0]}
        />
      </StoryItem>
    ))}
    </StoryWrapper>
  </Story>
</Preview>

### Radii

<Preview>
  <Story name="Radii">
    <StoryWrapper>
    {RADII.map(radius => (
      <StoryItem>
        <IconButton
          theme={THEMES[1]}
          radius={radius}
          Icon={Icons[0]}
        />
      </StoryItem>
    ))}
    </StoryWrapper>
  </Story>
</Preview>


### Loading

<Preview>
  <Story name="Loading">
    <StoryWrapper>
      <StoryItem>
        <IconButton
          isLoading
          Icon={Icons[0]}
        />
      </StoryItem>
      <StoryItem>
        <IconButton
          isDestructive
          isLoading
          Icon={Icons[0]}
        />
      </StoryItem>
    </StoryWrapper>
  </Story>
</Preview>