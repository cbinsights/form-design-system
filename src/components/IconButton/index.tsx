import React, { forwardRef } from 'react';
import cc from 'classcat';
import baseElement from 'util/baseElement';
import { IconProps } from 'components/Icon';

export const THEMES = ['ghost', 'aqua', 'outlined'];
export const RADII = ['square', 'circle'];
export const SIZES = ['s', 'm'];

export interface IconButtonProps {
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link?: React.ElementType;
  /** Controls active style UI of button */
  isActive?: boolean;
  /** Controls radius of button (slightly rounded square, or circle) */
  radius?: 'square' | 'circle';
  /** Controls spinner showing for icon button (normal icon is hidden) */
  isLoading?: boolean;
  /** Controls look and feel of button */
  theme?: 'ghost' | 'aqua' | 'outlined';
  /** Controls destructive look and feel of button */
  isDestructive?: boolean;
  /** Used to control the size of the button */
  size?: 's' | 'm';
  /** Controls whether the button is disabled or not. */
  disabled?: boolean;
  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon: React.ComponentType<IconProps>;
  /** Accessibility label */
  label: string;
  /** Extend click radius of button to nearest relative parent */
  isBreakoutLink?: boolean;
  /** URL for base element */
  href?: string;
  /** Handler for on click event */
  onClick?: () => void;
}

const IconButton = forwardRef(
  (
    {
      radius = 'square',
      theme = 'ghost',
      size = 'm',
      Icon,
      disabled,
      isActive,
      isLoading,
      isDestructive,
      label,
      Link,
      isBreakoutLink,
      ...rest
    }: IconButtonProps,
    ref
  ) => {
    const Element = baseElement({ href: rest.href, onClick: true, as: Link });
    return (
      <Element
        {...rest}
        ref={ref}
        title={label}
        className={cc([
          {
            'shape--circle': radius === 'circle',
            'fdsIconButton--disabled': disabled,
            'fdsIconButton--isActive': isActive && !disabled,
            'fdsIconButton--isDestructive': isDestructive,
            'fdsIconButton--loading': isLoading,
            breakoutLink: isBreakoutLink,
          },
          'fdsIconButton',
          'rounded--all',
          'border--focus--noTransition',
          'transition--default',
          'alignChild--center--center',
          `fdsIconButton--${theme}`,
          `fdsIconButton--${size}`,
        ])}
        disabled={disabled && Element === 'button'}
      >
        <div className="fdsIconButton--loading-spinnerWrapper">
          <div className={cc({ 'fdsIconButton--loading-spinner': isLoading })} />
        </div>
        <span className={isLoading ? 'fdsIconButton--hidden' : ''}>
          <Icon customSize={size === 's' ? 16 : 18} />
        </span>
      </Element>
    );
  }
);

IconButton.displayName = 'IconButton';

export default IconButton;
