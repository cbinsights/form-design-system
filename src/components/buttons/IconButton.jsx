import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from '../../util/baseElement';

export const THEMES = ['aqua', 'ghost'];
export const RADII = ['square', 'circle'];
export const SIZES = ['s', 'm'];

const IconButton = ({
  Icon,
  disabled,
  isActive,
  radius,
  theme,
  isLoading,
  size,
  isDestructive,
  label,
  Link,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, onClick: true, as: Link });
  return (
    <Element
      {...rest}
      title={label}
      className={cx(
        'fdsIconButton',
        'rounded--all',
        'border--focus--noTransition',
        'transition-default',
        `fdsIconButton--${theme}`,
        `fdsIconButton--${size}`,
        {
          'fdsIconButton--disabled': disabled,
          'fdsIconButton--active': isActive,
          'fdsIconButton--circle': radius === 'circle',
          'fdsIconButton--isDestructive': isDestructive,
          'fdsIconButton--loading': isLoading,
        }
      )}
      disabled={disabled && Element === 'button'}
    >
      <span className={isLoading ? 'fdsIconButton--hidden' : ''}>
        <Icon customSize={size === 's' ? 16 : 18} />
      </span>
    </Element>
  );
};

IconButton.defaultProps = {
  radius: 'square',
  theme: 'ghost',
  size: 'm',
};

IconButton.propTypes = {
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
  /** Controls active style UI of button */
  isActive: PropTypes.bool,
  /** Controls radius of button (slightly rounded square, or circle) */
  radius: PropTypes.oneOf(RADII),
  /** Controls spinner showing for icon button (normal icon is hidden) */
  isLoading: PropTypes.bool,
  /** Controls look and feel of button */
  theme: PropTypes.oneOf(THEMES),
  /** Controls destructive look and feel of button */
  isDestructive: PropTypes.bool,
  /** Used to control the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Controls whether the button is disabled or not. */
  disabled: PropTypes.bool,
  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon: PropTypes.func,
  /** Accessibility label */
  label: PropTypes.string,
};

export default IconButton;
