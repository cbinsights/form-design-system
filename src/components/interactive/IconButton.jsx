import React from 'react';
import PropTypes from 'prop-types';
import cc from 'classcat';
import baseElement from 'util/baseElement';

export const THEMES = ['ghost', 'aqua', 'outlined'];
export const RADII = ['square', 'circle'];
export const SIZES = ['s', 'm'];

const IconButton = ({
  radius = 'square',
  theme = 'ghost',
  size = 'm',
  Icon,
  disabled,
  isActive,
  isLoading,
  isDestructive,
  label,
  Link,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, onClick: true, as: Link });
  return (
    <Element
      {...rest}
      title={label}
      className={cc([
        {
          'shape--circle': radius === 'circle',
          'fdsIconButton--disabled': disabled,
          'fdsIconButton--isActive': isActive && !disabled,
          'fdsIconButton--isDestructive': isDestructive,
          'fdsIconButton--loading': isLoading,
        },
        'fdsIconButton',
        'rounded--all',
        'border--focus--noTransition',
        'transition--default',
        'alignChild--center--center',
        `fdsIconButton--${theme}`,
        `fdsIconButton--${size}`,
      ])}
      disabled={disabled && Element === 'button'}
    >
      <span className={isLoading ? 'fdsIconButton--hidden' : ''}>
        <Icon customSize={size === 's' ? 16 : 18} />
      </span>
    </Element>
  );
};

IconButton.propTypes = {
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
  /** Controls active style UI of button */
  isActive: PropTypes.bool,
  /** Controls radius of button (slightly rounded square, or circle) */
  radius: PropTypes.oneOf(RADII),
  /** Controls spinner showing for icon button (normal icon is hidden) */
  isLoading: PropTypes.bool,
  /** Controls look and feel of button */
  theme: PropTypes.oneOf(THEMES),
  /** Controls destructive look and feel of button */
  isDestructive: PropTypes.bool,
  /** Used to control the size of the button */
  size: PropTypes.oneOf(SIZES),
  /** Controls whether the button is disabled or not. */
  disabled: PropTypes.bool,
  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon: PropTypes.func.isRequired,
  /** Accessibility label */
  label: PropTypes.string.isRequired,
};

export default IconButton;
