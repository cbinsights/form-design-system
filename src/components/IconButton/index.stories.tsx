import React from 'react';
import StarFilledIcon from 'icons/react/StarFilledIcon';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';
import State from 'util/storybook-docs/State';

import IconButton, { THEMES, RADII, SIZES, IconButtonProps } from '.';

const Template = (args: IconButtonProps): JSX.Element => <IconButton {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  Icon: StarFilledIcon,
};

export const Themes = (args: IconButtonProps): JSX.Element => (
  <StoryWrapper>
    {THEMES.map((theme, idx) => (
      <StoryItem key={idx}>
        <IconButton {...args} {...Primary.args} theme={theme} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const Destructive = Template.bind({});
Destructive.args = {
  ...Primary.args,
  isDestructive: true,
};

export const Inverted = (args: IconButtonProps): JSX.Element => (
  <div className="inverted bgColor--blue padding--all rounded--all">
    <StoryWrapper>
      <StoryItem>
        <IconButton {...args} {...Primary.args} theme="ghost" />
      </StoryItem>
      <StoryItem>
        <IconButton {...args} {...Primary.args} theme="outlined" />
      </StoryItem>
    </StoryWrapper>
  </div>
);

export const IsActive = (args: IconButtonProps): JSX.Element => (
  <>
    <StoryWrapper>
      <State
        initialValue={{}}
        render={(value, setValue) => (
          <>
            {THEMES.map((theme, idx) => (
              <StoryItem key={idx}>
                <IconButton
                  {...args}
                  {...Primary.args}
                  theme={theme}
                  isActive={!value[idx]}
                  onClick={() => setValue({ ...value, [idx]: !value[idx] })}
                />
              </StoryItem>
            ))}
          </>
        )}
      />
      <State
        initialValue={false}
        render={(value, setValue) => (
          <>
            <StoryItem>
              <IconButton
                {...args}
                {...Primary.args}
                isDestructive
                isActive={!value[0]}
                onClick={() => setValue(!value)}
              />
            </StoryItem>
          </>
        )}
      />
    </StoryWrapper>
    <div className="inverted bgColor--blue padding--all rounded--all margin--top">
      <StoryWrapper>
        <State
          initialValue={{}}
          render={(value, setValue) => (
            <>
              {['ghost', 'outlined'].map((theme, idx) => (
                <StoryItem key={idx}>
                  <IconButton
                    {...args}
                    {...Primary.args}
                    theme={theme}
                    isActive={!value[idx]}
                    onClick={() => setValue({ ...value, [idx]: !value[idx] })}
                  />
                </StoryItem>
              ))}
            </>
          )}
        />
      </StoryWrapper>
    </div>
  </>
);

IsActive.parameters = {
  docs: {
    description: {
      story:
        '`isActive` should be used on buttons to signify the button having some affect (applying a filter, causing a popover to be open). **Click to toggle on and off**.',
    },
  },
};

export const Sizes = (args: IconButtonProps): JSX.Element => (
  <StoryWrapper>
    {SIZES.map((size, idx) => (
      <StoryItem key={idx}>
        <IconButton {...args} {...Primary.args} size={size} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const Radii = (args: IconButtonProps): JSX.Element => (
  <StoryWrapper>
    {RADII.map((radius, idx) => (
      <StoryItem key={idx}>
        <IconButton {...args} {...Primary.args} theme={THEMES[1]} radius={radius} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const Breakout = (args: IconButtonProps): JSX.Element => (
  <StoryWrapper>
    <StoryItem>
      <div
        className="alignChild--center--center padding--all--xl"
        style={{ border: '1px dashed #f09', position: 'relative' }}
      >
        <IconButton {...args} {...Primary.args} isBreakoutLink />
      </div>
    </StoryItem>
  </StoryWrapper>
);

Breakout.parameters = {
  docs: {
    description: {
      story: 'Breakout buttons extend their clickable area to the next nearest parent.',
    },
  },
};

export const Loading = (args: IconButtonProps): JSX.Element => (
  <StoryWrapper>
    <StoryItem>
      <IconButton {...args} {...Primary.args} isLoading />
    </StoryItem>
    <StoryItem>
      <IconButton {...args} {...Primary.args} isDestructive isLoading />
    </StoryItem>
  </StoryWrapper>
);

export default {
  component: IconButton,
  title: 'components/IconButton',
  parameters: {
    docs: {
      fds: {
        attributes: ['rest'],
      },
    },
  },
};
