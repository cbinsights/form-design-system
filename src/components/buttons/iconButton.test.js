import React from 'react';
import { shallow } from 'enzyme';

import IconButton from './IconButton';

const renderComponent = (props) => shallow(<IconButton {...props} />);

const props = {
  Icon: () => {}, isActive: true, radius: 'circle', size: 's', isLoading: true, theme: 'aqua', isDestructive: true, disabled: true
};

describe('ButtonGroup component', () => {

  it('matches snapshot (default props)', () => {
    const component = renderComponent({ Icon: () => {} });
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = renderComponent(props);
    expect(component).toMatchSnapshot();
  });
});
