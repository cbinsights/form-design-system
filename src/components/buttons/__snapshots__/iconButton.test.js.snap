// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup component matches snapshot (default props) 1`] = `
<button
  className="fdsIconButton rounded--all border--focus--noTransition transition-default fdsIconButton--ghost fdsIconButton--m"
>
  <span
    className=""
  >
    <Icon
      customSize={18}
    />
  </span>
</button>
`;

exports[`ButtonGroup component matches snapshot (set all props) 1`] = `
<button
  className="fdsIconButton rounded--all border--focus--noTransition transition-default fdsIconButton--aqua fdsIconButton--s fdsIconButton--disabled fdsIconButton--active fdsIconButton--circle fdsIconButton--isDestructive fdsIconButton--loading"
  disabled={true}
>
  <span
    className="fdsIconButton--hidden"
  >
    <Icon
      customSize={16}
    />
  </span>
</button>
`;
